### PR TITLE
Implement Squad Maker

### PR DESCRIPTION
**Pull request by Benobi42**
Working version is available at http://206.189.222.135/

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+.vscode/settings.json
```
Comment (gralamin) [inline .gitignore]: This looks like a file for specifically your editor. In general, these should not be added to `.gitignore`, but added to `.git/info/exclude` on your local repo.

File: .gitignore [hunk 2]
```diff
@@ -4,3 +4,48 @@ __pycache__/players.cpython-37.pyc
 __pycache__/squads.cpython-37.pyc
 templates/squad1.html
 templates/waitList.html
+test/.pytest_cache/v/cache/nodeids
+test/.pytest_cache/v/cache/lastfailed
+test/__pycache__/__init__.cpython-37.pyc
+test/__pycache__/test_players.cpython-37-PYTEST.pyc
+.pytest_cache/v/cache/nodeids
+test/__pycache__/test_squads.cpython-37-PYTEST.pyc
+templates/squad2.html
```
Comment (gralamin) [inline .gitignore]: I am confused, why not just `templates/squad*.html`?

File: .gitignore [hunk 3]
```diff
@@ -4,3 +4,48 @@ __pycache__/players.cpython-37.pyc
 __pycache__/squads.cpython-37.pyc
 templates/squad1.html
 templates/waitList.html
+test/.pytest_cache/v/cache/nodeids
+test/.pytest_cache/v/cache/lastfailed
+test/__pycache__/__init__.cpython-37.pyc
```
Comment (gralamin) [inline .gitignore]: Why not `*.pyc`?

File: templates/index.html [hunk 4]
```diff
@@ -0,0 +1,51 @@
+<!DOCTYPE html>
+<html lang="en">
+
+    <style>
```
Comment (gralamin) [inline templates/index.html:4]: I would put this in a `css` file in general, but it is likely fine to leave this in the template for now.

File: players.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
+import json
+import os
+
+from flask_table import Table, Col
+
+# Default JSON Data file
+JSONFILE = "players.json"
+
+# Default HTML file for Waiting List
+WAITLIST = "templates/waitList.html"
```
Comment (gralamin) [inline players.py]: You have referred to this file before it exists.

File: players.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+SKILLTYPES = ["Skating", "Shooting", "Checking"]
+
+
+def getSkillRating(sklData, sklType):
+    """
+    Get the rating for a given skill from
+    a single player's skill data
+
+    Skills are expected to be in one of the predetermined
+    skill categories
```
Comment (gralamin) [inline players.py:17]: If python 3, use enums, if python 2, you can get Enums, likely not worth it if this is the only enum.

File: players.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    """
+    if sklType not in SKILLTYPES:
+        raise ValueError("%s is not a valid skill")
+    return [skl for skl in sklData if skl['type'] == sklType][0]['rating']
+
+
+def writeHTMLToFile(html, fileName):
```
Comment (gralamin) [inline players.py]: Dynamically writing html files to use isn't a very maintainable solution. I suggest thinking carefully if this is what you want to do, instead of using templating systems such as jinja.

File: players.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        self._id = id
+        self.name = name
+        self.skate = skateRating
+        self.shoot = shootRating
+        self.check = checkRating
+
+    def __eq__(self, other):
+        """
+        Override the equality operator so that two players are
+        only equal if all of their attributes are equal.
```
Comment (gralamin) [inline players.py:46]: This docstring is currently a flat out lie.

File: players.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+
+# Default HTML file for Waiting List
+WAITLIST = "templates/waitList.html"
+
+SKILLTYPES = ["Skating", "Shooting", "Checking"]
+
+
+def getSkillRating(sklData, sklType):
+    """
+    Get the rating for a given skill from
```
Comment (gralamin) [inline players.py:13]: Docstrings don't appear to follow pep257. You haven't set that as a thing you use, but definitely consider setting up pydocstyle to run through your code via flake8.

File: test/test_players.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+
+class TestPlayer(TestCase):
+
+    def testCreatePlayer(self):
+        player = players.Player("123", "Ben Schreiber", 50, 50, 50)
+        self.assertTrue(isinstance(player, players.Player))
+
+    def testEquality(self):
```
Comment (gralamin) [inline test/test_players.py]: Also test the negative cases. You would notice your docstring is a lie if you did so.

File: test/test_players.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+        }
+    ]
+}
+
+
+class TestPlayer(TestCase):
+
+    def testCreatePlayer(self):
+        player = players.Player("123", "Ben Schreiber", 50, 50, 50)
+        self.assertTrue(isinstance(player, players.Player))
```
Comment (gralamin) [inline test/test_players.py]: ...How exactly would you imagine this not passing? Is this test actually adding anything?

File: players.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+
+    def toHTML(self, fn=WAITLIST):
+        """
+        Update and write out the HTML table corresponding to the list of
+        players.
+
+        Defaults to printing out to the HTML file for the Waiting List
+        """
+        self.table = PlayerTable(self.players, ["playerList"])
+        html = self.table.__html__()
```
Comment (gralamin) [inline players.py]: Generally it is bad style to use __html__... but this seems to be expected by flask-table's documentation. Nothing to do here, I guess.

File: players.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
+import json
+import os
+
+from flask_table import Table, Col
```
Comment (gralamin) [inline players.py]: At the point of this commit, your project is not set up with this requirement, or instructions on how to get it. I assume this is done later, but try to maintain a good state in every commit.

File: squads.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+        shootSum = 0
+        checkSum = 0
+        for player in self.players:
+            skateSum += player.skate
+            shootSum += player.shoot
+            checkSum += player.check
+
+        pLen = len(self.players)
+        return Player(None, AVERAGENAME, floor(skateSum/pLen),
+                      floor(shootSum/pLen), floor(checkSum/pLen))
```
Comment (gralamin) [inline squads.py]: You can also use integer division explicitly: `x//y`. In python, integer division uses floor (`-1 // 2 == -1`)

File: test/test_squads.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+from players import Player, PlayerList, PlayerTable
+
+
+class TestSquad(TestCase):
+    player1 = Player("123", "Ben Schreiber", 51, 51, 51)
+    player2 = Player("99", "Wayne Gretzky", 99, 99, 99)
+    playerList = PlayerList([player1, player2])
+
+    def testGetAveragePlayer(self):
```
Comment (gralamin) [inline test/test_squads.py]: A couple things worth checking you don't:
1. An odd number result of the sum.
2. Different values for different attributes, to make sure that a score isn't overwriting another.

File: templates/index.html [hunk 16]
```diff
@@ -26,26 +26,9 @@ <h1>Hockey Squad Builder</h1>
...
-            <tr>
-              <td>Eve Jackson</td>
-              <td>95</td> 
-              <td>95</td>
-              <td>95</td>
-            </tr>
-          </table>
+        <div>
+            <h3>Waiting List</h3>
+            {% include 'waitList.html' %}
```
Comment (gralamin) [inline templates/index.html]: Why not pass in the table from playerList and use ``{{table_variable_name}}``, instead of writing to html?

File: templates/index.html [hunk 17]
```diff
@@ -24,7 +24,7 @@ <h1>Hockey Squad Builder</h1>
 
         <button type="button" id=genSquads>Generate Squads</button> <br>
 
-        <button type="button" id=reset>Reset Squads</button>
+        <button type="button" id=reset onClick="window.location.reload()">Reset Squads</button>
```
Comment (gralamin) [inline templates/index.html]: This is fine for now, since you don't maintain any state, and don't need to maintain state. But this essentially leaves this feature "unimplemented" for when a real database etc. is behind it.

File: squads.py [hunk 18]
```diff
@@ -52,3 +52,114 @@ def toHTML(self):
...
+
+
+def tournamentRank(players):
+    """
+    Rank the players as if they were in a single elimintion tournament
+
+    This ranking is accomplished by sorting the players by the sum
+    of their three skills. The rankings are converted into their
+    binary representation with the same number of bits as the total
+    number of player, left padded with zeros, whjch are then reversed
```
Comment (gralamin) [inline squads.py]: Spelling error ("whjch")

File: squads.py [hunk 19]
```diff
@@ -52,3 +52,114 @@ def toHTML(self):
...
+    to generate their 'slot' in the tournament. Due to the nature of
+    this conversion, some numbers may be converted into invalid slots,
+    so the players are simply added to the tournament list in order
+    of increasing 'slot'.
+    """
+    tournament = []
```
Comment (gralamin) [inline squads.py]: I would suggest using a `deque`

File: squads.py [hunk 20]
```diff
@@ -52,3 +52,114 @@ def toHTML(self):
...
+                          "maximum value of %d" % (value, maxValue)))
+    bitLengthDiff = maxValue.bit_length()-value.bit_length()
+    bits = '0'*bitLengthDiff + bin(value)[2:]
+    reversedBits = bits[::-1]
+    return int(reversedBits, 2)
+
+
+def tournamentRank(players):
+    """
+    Rank the players as if they were in a single elimintion tournament
```
Comment (gralamin) [inline squads.py]: spelling ("elimintion")

File: squads.py [hunk 21]
```diff
@@ -52,3 +52,114 @@ def toHTML(self):
...
+    """
+    Dynamically generate psuedo balanced squads of players from a given list
+
+    Algorithm to balance teams is based on https://stackoverflow.com/a/1363601.
+    Players are ranked by the sum of their three skills, and this ranking is
+    used to generate their place in a single-elimination tournament. Once the
+    players are placed in this tournament order, each team alternates taking
+    the top and bottom players of the tournament into their team, until all
+    teams have reached the maximum number of players possible given the
+    desired number of teams. Remaining players are left on the waiting list.
```
Comment (gralamin) [inline squads.py]: This isn't really what your algorithm does, as far as I can tell.

Let's say we have players with the following seeds:

1
2
3
4
5
6

When you generate the tournament rank, you generate pairings like this:

4, 2
6, 1
5, 3

You then make this one list:

4, 2, 6, 1, 5, 3

From there, say you have two teams. The first team takes 4, the second team takes 2. The first team takes 3, the second team takes 5, the first team takes 6, the last team takes 1, generating:

4, 3, 6 = 13
2, 5, 1 = 8
Here you have the best two players on a team with the SECOND WORST player, and the WORST player with the middle players.

A more balanced team would be taken using your alternate choosing with a single list of sums of skills, that would give you:

1, 6, 3 = 10
2, 5, 4 = 11

These look much fairer teams then your current algorithm.

File: test/test_squads.py [hunk 22]
```diff
@@ -34,3 +34,164 @@ def testToHTML(self):
...
+    def testGetBalancedSquads(self):
+        numSquads = 2
+        player1 = Player("99", "Wayne Gretzky", 99, 99, 99)
+        player2 = Player("97", "Connor McDavid", 97, 97, 97)
+        player3 = Player("123", "Ben Schreiber", 51, 51, 51)
+        player4 = Player("404", "Nota RealPlayer", 20, 20, 20)
+        playerList = PlayerList([player1, player2, player3, player4])
+
+        expectedSquads = [squads.Squad(1, [player4, player3]),
+                          squads.Squad(2, [player2, player1])]
```
Comment (gralamin) [inline test/test_squads.py]: Just looking at this test case should point out the problems I mentioned above: These squads aren't balanced AT ALL

File: app.py [hunk 23]
```diff
@@ -16,8 +19,26 @@ def main():
...
+@app.route("/squads", methods=["POST"])
+def generateSquads():
+    """
+    Generate a number of balanced squads equivalent to the user's input from
+    the given player data, and enable them to be displayed on the new page.
+    """
+    numSquads = int(request.form["numSquads"])
+    playerList = PlayerList.fromJSON()
+    squads = getBalancedSquads(numSquads, playerList)
+    squadsHTML = [("squad%d.html" % (i+1)) for i in range(numSquads)]
```
Comment (gralamin) [inline app.py]: As before, this "include" choice is pretty unnecessary seeming...

File: squads.py [hunk 24]
```diff
@@ -103,10 +103,10 @@ def getBalancedSquads(numSquads, playerList):
...
     Players are ranked by the sum of their three skills, and this ranking is
     used to generate their place in a single-elimination tournament. Once the
-    players are placed in this tournament order, each team alternates taking
-    the top and bottom players of the tournament into their team, until all
-    teams have reached the maximum number of players possible given the
-    desired number of teams. Remaining players are left on the waiting list.
+    players are placed in this tournament order, each squad, starting with the
+    first, takes the next x players, where x is the desired squad size needed
+    to generate the number of squads desired. Remaining players are left on the
+    waiting list.
```
Comment (gralamin) [inline squads.py]: Okay, so you have completely rewritten the algorithm here. That is good, but it does leave your reviewer in a state of "wait, why are you using this terrible algorithm?".

File: squads.py [hunk 25]
```diff
@@ -41,12 +41,18 @@ def getAveragePlayer(self):
...
+        squad number, and the average of all skills on the team at the
+        bottom.
         """
-        self.players.append(self.getAveragePlayer())
+        avgPlay = self.getAveragePlayer()
         self.table = PlayerTable(self.players, ["playerList"])
         html = self.table.__html__()
-        self.players.pop()
+
+        html = html.replace("</tbody>",
```
Comment (gralamin) [inline squads.py]: Having to do this in place really suggests you aren't doing what you want to be doing...

File: templates/index.html [hunk 26]
```diff
@@ -2,40 +2,141 @@
 <html lang="en">
 
     <style>
-        table, th, td {
-            border: 1px solid black;
-            border-collapse: collapse;
+        body {
```
Comment (gralamin) [inline templates/index.html]: This is a ton of UI updates! I feel you should split this commit up into a few chunks, the UI isn't something to ignore, clearly articulate why you are making each update and what you are accomplishing.

**Review by gralamin — COMMENTED** [review]
A few things I think are definitely missing not addressed in comments below:

1. You should set up pylint / flake8 / etc. to be able to run against your code.
2. You should set up a script to run your unit tests.

I suspect there are a number of style problems in the code that are being ignored currently, specifically around spaces and operators, and docstrings.

File: .gitignore (hunk 2 above)
Comment (Benobi42) [reply .gitignore]: Because I had visual studio code auto-gitignore the files, will change.

File: players.py (hunk 5 above)
Comment (gralamin) [reply players.py]: In retrospect, this is fine, just make it clearer this file is generating this file, my expectation is it was using an existing file.

File: squads.py (hunk 21 above)
Comment (gralamin) [reply squads.py]: Ben and I talked about this, with regards to his later changes and wanting to keep showing the work. My suggestion was to tell me that this will change in his commit message.

File: players.py (hunk 7 above)
Comment (Benobi42) [reply players.py]: You are entirely correct, I will change it to use actual jinja templating instead of faking it

File: test/test_squads.py (hunk 15 above)
Comment (Benobi42) [reply test/test_squads.py]: Good ideas, adding tests for those

File: squads.py (hunk 14 above)
Comment (Benobi42) [reply squads.py]: Right, switching to that

File: templates/index.html (hunk 16 above)
Comment (Benobi42) [reply templates/index.html]: Your suggestion is the correct way to do it, changing

File: players.py (hunk 13 above)
Comment (Benobi42) [reply players.py]: Didn't think of the @requirements.txt until dockerization, but I will move it back to follow along with requirements

File: players.py (hunk 8 above)
Comment (Benobi42) [reply players.py:46]: Fixed to use the intended comparison

File: test/test_players.py (hunk 11 above)
Comment (Benobi42) [reply test/test_players.py]: ...no. Removing

File: test/test_players.py (hunk 10 above)
Comment (Benobi42) [reply test/test_players.py]: Good idea, testing now

File: squads.py (hunk 21 above)
Comment (Benobi42) [reply squads.py]: On further consideration, I am moving the later algorithm to the same commit, and will investigate on using your method here
